File: wrappers/python/aries_askar/ecdh.py
from typing import Optional, Union

from .bindings import Encrypted, key_derive_ecdh_es, key_derive_ecdh_1pu
from .key import Key
from .types import KeyAlg


def _load_key(key: Union[dict, str, Key]) -> Key:
    if isinstance(key, (str, dict)):
        key = Key.from_jwk(key)
    return key


class EcdhEs:
    def __init__(
        self,
        alg_id: Union[bytes, str],
        apu: Union[bytes, str],
        apv: Union[bytes, str],
    ):
        self._params = (alg_id, apu, apv)

    def _derive_key(
        self,
        enc_alg: Union[str, KeyAlg],
        ephemeral_key: Key,
        receiver_key: Key,
        receive: bool,
    ) -> Key:
        return Key(
            key_derive_ecdh_es(
                enc_alg,
                ephemeral_key._handle,
                receiver_key._handle,
                *self._params,
                receive,
            )
        )

    def encrypt_direct(
        self,
        enc_alg: Union[str, KeyAlg],
        ephemeral_key: Union[dict, str, Key],
        receiver_key: Union[dict, str, Key],
        message: Union[str, bytes],
        *,
        aad: bytes = None,
        nonce: bytes = None,
    ) -> Encrypted:
        derived = self._derive_key(
            enc_alg, _load_key(ephemeral_key), _load_key(receiver_key), False
        )
        return derived.aead_encrypt(message, nonce=nonce, aad=aad)

    def decrypt_direct(
        self,
        enc_alg: Union[str, KeyAlg],
        ephemeral_key: Union[dict, str, Key],
        receiver_key: Union[dict, str, Key],
        ciphertext: bytes,
        *,
        nonce: bytes,
        tag: bytes,
        aad: bytes = None,
    ) -> bytes:
        derived = self._derive_key(
            enc_alg, _load_key(ephemeral_key), _load_key(receiver_key), True
        )
        return derived.aead_decrypt(ciphertext, nonce=nonce, tag=tag, aad=aad)

    def sender_wrap_key(
        self,
        wrap_alg: Union[str, KeyAlg],
        ephemeral_key: Union[dict, str, Key],
        receiver_key: Union[dict, str, Key],
        cek: Key,
    ) -> Encrypted:
        derived = self._derive_key(
            wrap_alg, _load_key(ephemeral_key), _load_key(receiver_key), False
        )
        return derived.wrap_key(cek)

    def receiver_unwrap_key(
        self,
        wrap_alg: Union[str, KeyAlg],
        enc_alg: Union[str, KeyAlg],
        ephemeral_key: Union[dict, str, Key],
        receiver_key: Union[dict, str, Key],
        ciphertext: bytes,
        *,
        nonce: bytes = None,
        tag: bytes = None,
    ) -> Key:
        derived = self._derive_key(
            wrap_alg, _load_key(ephemeral_key), _load_key(receiver_key), True
        )
        return derived.unwrap_key(enc_alg, ciphertext, nonce=nonce, tag=tag)


class Ecdh1PU:
    def __init__(
        self,
        alg_id: Union[bytes, str],
        apu: Union[bytes, str],
        apv: Union[bytes, str],
    ):
        self._params = (alg_id, apu, apv)

    def _derive_key(
        self,
        key_alg: Union[str, KeyAlg],
        ephemeral_key: Key,
        sender_key: Key,
        receiver_key: Key,
        cc_tag: Optional[bytes],
        receive: bool,
    ) -> Key:
        return Key(
            key_derive_ecdh_1pu(
                key_alg,
                ephemeral_key._handle,
                sender_key._handle,
                receiver_key._handle,
                *self._params,
                cc_tag,
                receive,
            )
        )

    def encrypt_direct(
        self,
        key_alg: Union[str, KeyAlg],
        ephemeral_key: Union[dict, str, Key],
        sender_key: Union[dict, str, Key],
        receiver_key: Union[dict, str, Key],
        message: Union[str, bytes],
        *,
        aad: bytes = None,
        nonce: bytes = None,
    ) -> Encrypted:
        derived = self._derive_key(
            key_alg,
            _load_key(ephemeral_key),
            _load_key(sender_key),
            _load_key(receiver_key),
            None,
            False,
        )
        return derived.aead_encrypt(message, nonce=nonce, aad=aad)

    def decrypt_direct(
        self,
        enc_alg: Union[str, KeyAlg],
        ephemeral_key: Union[dict, str, Key],
        sender_key: Union[dict, str, Key],
        receiver_key: Union[dict, str, Key],
        ciphertext: bytes,
        *,
        nonce: bytes,
        tag: bytes,
        aad: bytes = None,
    ) -> bytes:
        derived = self._derive_key(
            enc_alg,
            _load_key(ephemeral_key),
            _load_key(sender_key),
            _load_key(receiver_key),
            None,
            True,
        )
        return derived.aead_decrypt(ciphertext, nonce=nonce, tag=tag, aad=aad)

    def sender_wrap_key(
        self,
        wrap_alg: Union[str, KeyAlg],
        ephemeral_key: Union[dict, str, Key],
        sender_key: Union[dict, str, Key],
        receiver_key: Union[dict, str, Key],
        cek: Key,
        *,
        cc_tag: bytes,
    ) -> Encrypted:
        derived = self._derive_key(
            wrap_alg,
            _load_key(ephemeral_key),
            _load_key(sender_key),
            _load_key(receiver_key),
            cc_tag=cc_tag,
            receive=False,
        )
        return derived.wrap_key(cek)

    def receiver_unwrap_key(
        self,
        wrap_alg: Union[str, KeyAlg],
        enc_alg: Union[str, KeyAlg],
        ephemeral_key: Union[dict, str, Key],
        sender_key: Union[dict, str, Key],
        receiver_key: Union[dict, str, Key],
        ciphertext: bytes,
        *,
        cc_tag: bytes,
        nonce: bytes = None,
        tag: bytes = None,
    ) -> Key:
        derived = self._derive_key(
            wrap_alg,
            _load_key(ephemeral_key),
            _load_key(sender_key),
            _load_key(receiver_key),
            cc_tag=cc_tag,
            receive=True,
        )
        return derived.unwrap_key(enc_alg, ciphertext, nonce=nonce, tag=tag)

File: wrappers/swift/Askar/Sources/Askar/Ecdh.swift
import Foundation
import AskarFramework

public class EcdhEs {
    let algId: FfiByteBuffer
    let apu: FfiByteBuffer
    let apv: FfiByteBuffer

    public init(algId: String, apu: String? = nil, apv: String? = nil) {
        self.algId = FfiByteBuffer(fromString: algId)
        self.apu = FfiByteBuffer(fromString: apu)
        self.apv = FfiByteBuffer(fromString: apv)
    }

    func deriveKey(encAlg: KeyAlg, ephemeralKey: Key, receiverKey: Key, receive: Bool) throws -> Key {
        var handle = LocalKeyHandle()
        let error = askar_key_derive_ecdh_es(encAlg.rawValue, ephemeralKey.handle, receiverKey.handle, algId.buffer, apu.buffer, apv.buffer, receive ? 1:0, &handle)
        if error != Success {
            throw AskarError.nativeError(code: error.rawValue)
        }

        return Key(handle: handle)
    }

    public func encryptDirect(encAlg: KeyAlg,
                              ephemeralKey: Key,
                              receiverKey: Key,
                              message: Data,
                              aad: Data? = nil,
                              nonce: Data? = nil) throws -> Encrypted {
        let derived = try deriveKey(encAlg: encAlg, ephemeralKey: ephemeralKey, receiverKey: receiverKey, receive: false)
        return try derived.aeadEncrypt(message: message, nonce: nonce, aad: aad)
    }

    public func decryptDirect(encAlg: KeyAlg,
                              ephemeralKey: Key,
                              receiverKey: Key,
                              ciphertext: Data,
                              nonce: Data,
                              tag: Data,
                              aad: Data? = nil) throws -> Data {
        let derived = try deriveKey(encAlg: encAlg, ephemeralKey: ephemeralKey, receiverKey: receiverKey, receive: true)
        return try derived.aeadDecrypt(ciphertext: ciphertext, nonce: nonce, tag: tag, aad: aad)
    }

    public func senderWrapKey(wrapAlg: KeyAlg,
                              ephemeralKey: Key,
                              receiverKey: Key,
                              cek: Key) throws -> Encrypted {
        let derived = try deriveKey(encAlg: wrapAlg, ephemeralKey: ephemeralKey, receiverKey: receiverKey, receive: false)
        return try derived.wrapKey(other: cek)
    }

    public func receiverUnwrapKey(wrapAlg: KeyAlg,
                                  encAlg: KeyAlg,
                                  ephemeralKey: Key,
                                  receiverKey: Key,
                                  ciphertext: Data,
                                  nonce: Data? = nil,
                                  tag: Data? = nil) throws -> Key {
        let derived = try deriveKey(encAlg: wrapAlg, ephemeralKey: ephemeralKey, receiverKey: receiverKey, receive: true)
        return try derived.unwrapKey(alg: encAlg, ciphertext: ciphertext, nonce: nonce, tag: tag)
    }
}

public class Ecdh1PU {
    let algId: FfiByteBuffer
    let apu: FfiByteBuffer
    let apv: FfiByteBuffer

    public init(algId: String, apu: String? = nil, apv: String? = nil) {
        self.algId = FfiByteBuffer(fromString: algId)
        self.apu = FfiByteBuffer(fromString: apu)
        self.apv = FfiByteBuffer(fromString: apv)
    }

    func deriveKey(encAlg: KeyAlg, ephemeralKey: Key, senderKey: Key, receiverKey: Key, ccTag: Data? = nil, receive: Bool) throws -> Key {
        var handle = LocalKeyHandle()
        let ccTagBuf = FfiByteBuffer(fromData: ccTag)
        let error = askar_key_derive_ecdh_1pu(encAlg.rawValue, ephemeralKey.handle, senderKey.handle, receiverKey.handle, algId.buffer, apu.buffer, apv.buffer, ccTagBuf.buffer, receive ? 1:0, &handle)
        if error != Success {
            throw AskarError.nativeError(code: error.rawValue)
        }

        return Key(handle: handle)
    }

    public func encryptDirect(encAlg: KeyAlg,
                              ephemeralKey: Key,
                              senderKey: Key,
                              receiverKey: Key,
                              message: Data,
                              aad: Data? = nil,
                              nonce: Data? = nil) throws -> Encrypted {
        let derived = try deriveKey(encAlg: encAlg, ephemeralKey: ephemeralKey, senderKey: senderKey, receiverKey: receiverKey, receive: false)
        return try derived.aeadEncrypt(message: message, nonce: nonce, aad: aad)
    }

    public func decryptDirect(encAlg: KeyAlg,
                              ephemeralKey: Key,
                              senderKey: Key,
                              receiverKey: Key,
                              ciphertext: Data,
                              nonce: Data,
                              tag: Data,
                              aad: Data? = nil) throws -> Data {
        let derived = try deriveKey(encAlg: encAlg, ephemeralKey: ephemeralKey, senderKey: senderKey, receiverKey: receiverKey, receive: true)
        return try derived.aeadDecrypt(ciphertext: ciphertext, nonce: nonce, tag: tag, aad: aad)
    }

    public func senderWrapKey(wrapAlg: KeyAlg,
                              ephemeralKey: Key,
                              senderKey: Key,
                              receiverKey: Key,
                              cek: Key,
                              ccTag: Data) throws -> Encrypted {
        let derived = try deriveKey(encAlg: wrapAlg, ephemeralKey: ephemeralKey, senderKey: senderKey, receiverKey: receiverKey, ccTag: ccTag, receive: false)
        return try derived.wrapKey(other: cek)
    }

    public func receiverUnwrapKey(wrapAlg: KeyAlg,
                                  encAlg: KeyAlg,
                                  ephemeralKey: Key,
                                  senderKey: Key,
                                  receiverKey: Key,
                                  ciphertext: Data,
                                  ccTag: Data,
                                  nonce: Data? = nil,
                                  tag: Data? = nil) throws -> Key {
        let derived = try deriveKey(encAlg: encAlg, ephemeralKey: ephemeralKey, senderKey: senderKey, receiverKey: receiverKey, ccTag: ccTag, receive: true)
        return try derived.unwrapKey(alg: wrapAlg, ciphertext: ciphertext, nonce: nonce, tag: tag)
    }
}
